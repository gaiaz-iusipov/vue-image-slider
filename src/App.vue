<template>
  <div
    class="image-slider"
    v-if="!empty"
  >
    <transition name="fade" mode="in-out">
      <div
        class="image-slider__slide"
        v-bind:key="active"
        v-bind:style="`background-image: url('${images[active]}');`"
        v-if="active !== null"
      />
    </transition>
    <div class="image-slider__content" v-if="hasDefaultSlot" v-once>
      <slot />
    </div>
    <div
      class="image-slider__overlay"
      v-on:click="show(next)"
      v-on:mouseover="stop"
      v-on:mouseleave="start"
      v-once
    />
    <div class="image-slider__footer ">
      <div
        class="image-slider__indicator"
        v-bind:class="{
          'image-slider__indicator_active': indicator.active,
          'image-slider__indicator_inactive': !indicator.active,
          'image-slider__indicator_loading': indicator.loading
        }"
        v-for="(indicator, index) in indicators"
        v-bind:key="index"
        v-on:click.stop="show(index)"
        v-if="hasIndicatorsSlot"
      >
        <slot name="indicator" v-bind:indicator="indicator" />
      </div>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    value: {
      type: Array,
      required: true
    },
    delay: {
      type: Number,
      default () { return 3000 }
    }
  },
  data () {
    return {
      images: this.value || [],
      active: null,
      loading: null
    }
  },
  watch: {
    value (value) {
      this.images = value || []
    },
    images (value) {
      if (value.length === 1) {
        this.stop()
      }
    }
  },
  computed: {
    empty () {
      return this.images.length === 0
    },
    prev () {
      if (this.images.length < 2) {
        return null
      }

      if (this.active === 0) {
        return this.last
      }

      return this.active - 1
    },
    next () {
      if (this.images.length < 2) {
        return null
      }

      if (this.active === this.last) {
        return 0
      }

      return this.active + 1
    },
    last () {
      if (this.empty) {
        return null
      }

      return this.images.length - 1
    },
    hasDefaultSlot () {
      return !!this.$slots['default']
    },
    hasIndicatorsSlot () {
      return !!this.$scopedSlots['indicator']
    },
    indicators () {
      return this.images.map((image, index) => ({
        active: index === this.active,
        loading: index === this.loading
      }))
    }
  },
  created () {
    this.interval = null
  },
  mounted () {
    this.show(0)
    this.start()
  },
  methods: {
    start () {
      if (this.interval !== null) {
        return
      }

      if (this.images.length < 2) {
        return
      }

      this.interval = window.setInterval(() => this.show(this.next), this.delay)
    },
    stop () {
      if (this.interval === null) {
        return
      }

      window.clearInterval(this.interval)
      this.interval = null
    },
    async show (index) {
      if (this.active === index) {
        return
      }

      try {
        await this.loadImage(index)
      } catch (error) {
        this.remove(index)
        this.show(this.next)
        return
      }

      this.active = index
    },
    remove (index) {
      this.images.splice(index, 1)
      this.$emit('input', this.images)
    },
    loadImage (index) {
      const url = this.images[index]

      return new Promise((resolve, reject) => {
        const image = new Image()
        image.src = url

        if (image.compleate || image.naturalHeight !== 0) {
          resolve()
          return
        }

        this.stop()
        this.loading = index

        image.addEventListener('load', event => {
          this.loading = null
          this.start()
          resolve()
        })
        image.addEventListener('error', event => {
          this.start()
          reject(new Error(`Failed to load image's URL: ${url}`))
        })
      })
    }
  }
}
</script>

<style scoped>
.image-slider {
  position: relative;
}
.image-slider__slide,
.image-slider__content,
.image-slider__overlay {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}
.image-slider__slide {
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center;
}
.image-slider__footer {
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  justify-content: center;
}
.image-slider__indicator {
  margin: .5rem .2rem;
  cursor: pointer;
}
</style>
