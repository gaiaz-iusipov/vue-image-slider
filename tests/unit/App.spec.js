import { shallowMount } from '@vue/test-utils'
import App from '@/App.vue'

const images = [
  'https://dummyimage.com/100&text=1',
  'https://dummyimage.com/300&text=2',
  'https://dummyimage.com/300&text=3'
]

const contentSlot = {
  default: 'content'
}

const indicatorSlot = {
  indicator: '<div />'
}

const factory = (images = [], slots = {}, scopedSlots = {}) => {
  return shallowMount(App, {
    propsData: { value: images },
    slots,
    scopedSlots
  })
}

describe('snapshots', () => {
  it('empty', () => {
    const wrapper = factory()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('inactive', () => {
    const wrapper = factory(images)

    expect(wrapper.element).toMatchSnapshot()
  })

  it('with content', () => {
    const wrapper = factory(images, contentSlot)

    expect(wrapper.element).toMatchSnapshot()
  })

  it('with images', () => {
    const wrapper = factory(images)
    wrapper.setData({ active: 0 })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('with images and indicators', () => {
    const wrapper = factory(images, {}, indicatorSlot)
    wrapper.setData({ active: 0 })

    expect(wrapper.element).toMatchSnapshot()
  })
})

describe('props', () => {
  it('value is required', () => {
    const wrapper = factory()
    const prop = wrapper.vm.$options.props.value

    expect(prop.required).toBeTruthy()
  })

  it('value is of type array', () => {
    const wrapper = factory()
    const prop = wrapper.vm.$options.props.value

    expect(prop.type).toBe(Array)
  })
})

describe('data', () => {
  it('images equal value', () => {
    const wrapper = factory(images)

    expect(wrapper.vm.images).toEqual(images)
  })
})

describe('computed properties', () => {
  it('prev equals null', () => {
    const wrapper = factory([images[0]])

    expect(wrapper.vm.prev).toBeNull()
  })

  it('prev equals last index', () => {
    const wrapper = factory(images)
    wrapper.setData({ active: 0 })

    expect(wrapper.vm.prev).toEqual(images.length - 1)
  })

  it('prev equals previous index', () => {
    const wrapper = factory(images)
    wrapper.setData({ active: 1 })

    expect(wrapper.vm.prev).toEqual(0)
  })

  it('next equals null', () => {
    const wrapper = factory([images[0]])

    expect(wrapper.vm.next).toBeNull()
  })

  it('next equals first index', () => {
    const wrapper = factory(images)
    wrapper.setData({ active: images.length - 1 })

    expect(wrapper.vm.next).toEqual(0)
  })

  it('next equals next index', () => {
    const wrapper = factory(images)
    wrapper.setData({ active: 0 })

    expect(wrapper.vm.next).toEqual(1)
  })

  it('last equals null', () => {
    const wrapper = factory()

    expect(wrapper.vm.last).toBeNull()
  })

  it('last equals last index', () => {
    const wrapper = factory(images)

    expect(wrapper.vm.last).toEqual(images.length - 1)
  })

  it('hasDefaultSlot equals false', () => {
    const wrapper = factory(images)

    expect(wrapper.vm.hasDefaultSlot).toBeFalsy()
  })

  it('hasIndicators equals true', () => {
    const wrapper = factory(images, contentSlot)

    expect(wrapper.vm.hasDefaultSlot).toBeTruthy()
  })

  it('hasIndicators equals false', () => {
    const wrapper = factory(images)

    expect(wrapper.vm.hasIndicatorsSlot).toBeFalsy()
  })

  it('hasIndicators equals true', () => {
    const wrapper = factory(images, {}, indicatorSlot)

    expect(wrapper.vm.hasIndicatorsSlot).toBeTruthy()
  })

  it('indicators has same size of images', () => {
    const wrapper = factory(images, {}, indicatorSlot)

    expect(wrapper.vm.indicators).toHaveLength(images.length)
  })
})

describe('rendering', () => {
  it('empty element', () => {
    const wrapper = factory()

    expect(wrapper.isEmpty()).toBeTruthy()
  })

  it('slide element is not exists', () => {
    const wrapper = factory(images)

    expect(wrapper.find('.image-slider__slide').exists()).toBeFalsy()
  })

  it('slide element is exists', () => {
    const wrapper = factory(images)
    wrapper.setData({ active: 0 })

    expect(wrapper.find('.image-slider__slide').exists()).toBeTruthy()
  })

  it('slide element is empty', () => {
    const wrapper = factory(images)
    wrapper.setData({ active: 0 })

    expect(wrapper.find('.image-slider__slide').isEmpty()).toBeTruthy()
  })

  it('content element is not exists', () => {
    const wrapper = factory(images)

    expect(wrapper.find('.image-slider__content').exists()).toBeFalsy()
  })

  it('content element is exists', () => {
    const wrapper = factory(images, contentSlot)

    expect(wrapper.find('.image-slider__content').exists()).toBeTruthy()
  })

  it('content element equals slot', () => {
    const wrapper = factory(images, contentSlot)

    expect(wrapper.find('.image-slider__content').text()).toEqual(contentSlot.default)
  })

  it('overlay element is exists', () => {
    const wrapper = factory(images)

    expect(wrapper.find('.image-slider__overlay').exists()).toBeTruthy()
  })

  it('footer element is exists', () => {
    const wrapper = factory(images)

    expect(wrapper.find('.image-slider__footer').exists()).toBeTruthy()
  })

  it('footer element is empty', () => {
    const wrapper = factory(images)

    expect(wrapper.find('.image-slider__footer').isEmpty()).toBeTruthy()
  })

  it('indicator elements are exist', () => {
    const wrapper = factory(images, {}, indicatorSlot)

    expect(wrapper.findAll('.image-slider__indicator')).toHaveLength(images.length)
  })
})
