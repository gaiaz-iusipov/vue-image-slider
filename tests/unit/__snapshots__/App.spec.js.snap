// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots empty 1`] = `<!---->`;

exports[`snapshots inactive 1`] = `
<div
  class="image-slider"
>
  <!---->
   
  <div
    class="image-slider__overlay"
  />
   
  <div
    class="image-slider__footer "
  >
    <!---->
    <!---->
    <!---->
  </div>
</div>
`;

exports[`snapshots with images 1`] = `
<div
  class="image-slider"
>
  <div
    class="image-slider__slide"
    mode="in-out"
    name="fade"
    style="background-image: url(https://dummyimage.com/100&text=1);"
  />
   
  <div
    class="image-slider__overlay"
  />
   
  <div
    class="image-slider__footer "
  >
    <!---->
    <!---->
    <!---->
  </div>
</div>
`;

exports[`snapshots with images and indicators 1`] = `
<div
  class="image-slider"
>
  <div
    class="image-slider__slide"
    mode="in-out"
    name="fade"
    style="background-image: url(https://dummyimage.com/100&text=1);"
  />
   
  <div
    class="image-slider__overlay"
  />
   
  <div
    class="image-slider__footer "
  >
    <div
      class="image-slider__indicator image-slider__indicator_active image-slider__indicator_loading"
    >
      <div />
    </div>
    <div
      class="image-slider__indicator image-slider__indicator_inactive"
    >
      <div />
    </div>
    <div
      class="image-slider__indicator image-slider__indicator_inactive"
    >
      <div />
    </div>
  </div>
</div>
`;
